app/controllers/user_preferences_controller.rb

```diff
@@ -0,0 +1,53 @@
+class UserPreferencesController < ApplicationController
+  # User preferences is currently gated on bulk downloads because that's the only user preference
+  # and it doesn't make sense to show an empty screen.
+  # Modify this condition if you are adding an additional user preference and bulk downloads
+  # is not yet released.
+  before_action do
+    allowed_feature_required("bulk_downloads", true)
+  end
+
+  # GET /user_preferences/metadata_by_category
+  # Get metadata for user preferences that are editable by the user.
+  def metadata_by_category
+    user_preference_metadata = []
+    viewable_keys = current_user.viewable_user_preference_keys
+
+    UserPreference::DISPLAY_CATEGORIES.each do |category|
+      viewable_category_keys = category[:preferences] & viewable_keys
+
+      unless viewable_category_keys.empty?
+        user_preference_metadata << {
+          name: category[:name],
+          preferences: viewable_category_keys.map do |key|
+            {
+              key: key,
+              description: UserPreference::USER_PREFERENCE_METADATA[key][:description],
+              data_type: UserPreference::USER_PREFERENCE_METADATA[key][:data_type],
+            }
+          end,
+        }
+      end
+    end
+
+    render json: user_preference_metadata
+  end
+
+  # POST /user_preferences/update
+  def update
+    current_user.set_user_preference(params[:key], params[:value])
+
+    render json: {
+      status: "success",
+      key: params[:key],
+      value: current_user.get_user_preference(params[:key]),
+    }
+  rescue
+    render json: {
```

Comment:
as a matter of convention, it would be good to know whether we should always log http errors

Reply:
I think we should. Thanks for pointing it out here.